[ENH] Initial commit of dockerfile, updated docs

emdupre opened the pull request:
References #125 

Creating a dockerfile to serve jupyter-books, updating documentation to reference include the correct docker commands.

jasmainak commented on jupyter_book/book_template/content/guide/03_build.md:
````diff
@@ -45,13 +45,38 @@ From here, you have **two options**
+```
+
+And that's it! If you navigate to `http://0.0.0.0:4000/jupyter-book/` in your browser,
+you should see a preview copy of your book.
````
Neat!

emdupre replied:
Hopefully it's useful !!

jasmainak commented on jupyter_book/book_template/content/guide/03_build.md:
```diff
@@ -45,13 +45,38 @@ From here, you have **two options**
-steps to do so.
+If you'd like to build your book's site HTML locally, you can do so either with a
+"bare metal" installation (i.e., where you install all of the necessary dependencies yourself),
+or by using our provided [docker]() image.
```
link is missing?

emdupre replied:
Fix :) Thanks for catching !

jasmainak commented on jupyter_book/book_template/content/guide/03_build.md:
````diff
@@ -42,16 +42,52 @@ From here, you have **two options**
+
+```bash
+docker run --rm --security-opt label:disable  \
+   -v /PATH/TO/YOUR/CONTENT:/srv/jekyll \
````
does this need to be the full path? Because if I try `.` when I am in the current folder, it tells me:

```sh
docker: Error response from daemon: create .: volume name is too short, names should be at least two alphanumeric characters.
```

and when I try the full path, it tells me:

```sh
Could not find public_suffix-2.0.5 in any of the sources
Run `bundle install` to install missing gems.
```

emdupre replied:
It does need to be the full path ! Or you can use the `$PWD` variable.

And in my experience that means you need to delete your `Gemfile.lock`. Are you using the one in this PR ?

jasmainak replied:
Indeed, I had forgotten to copy the `Gemfile.lock` from this PR. It worked. Thanks :)

jasmainak replied:
So practically speaking, one would need to first upgrade their jupyter-book to use this, no?

emdupre replied:
I think we need to cut a new release soon because of #120 so hopefully this PR will make it in !

Or you _should_ be able to delete your local `Gemfile.lock`, copy in the `Dockerfile`, and run this command. If you want to try, would love feedback on it !

jasmainak replied:
Correct. But even then, existing books cannot use this unless they upgrade (?) Or is there a check somewhere so that the books are upgraded also when one starts using a new version?

emdupre replied:
Sorry, I think I'm missing something. Do you mean the new versions listed in the `Gemfile.lock` or a new version of Jupyter Book itself ?

jasmainak replied:
So, if I understand correctly, there are two copies of `Gemfile.lock` -- one in the jupyter-book repository itself and the second which is in the book I created, say my-book. Thus you have both jupyter_book/book_template/Gemfile.lock and my-book/Gemfile.lock. So, even if jupyter_book/book_template/Gemfile.lock is upgraded, my-book/Gemfile.lock will not be updated unless you run the command "jupyter-book upgrade my-book"

emdupre replied:
Ah, ok, thanks for clarifying ! Yes, because your book was made with an older version of Jupyter Book you'll need to update the files in them. I don't think it's worth it to implement an "upgrade" command, at least for now (if the changes become more substantial in the future then it's worth thinking about !). But I can add to the FAQ that if you see the above error, to try deleting your local `Gemfile.lock` ?

jasmainak replied:
Oh but you don't have to implement anything ! The feature [already exists](https://jupyter.org/jupyter-book/guide/04_faq.html). I was just suggesting adding a note above.

emdupre replied:
:tada: nice ! Sorry, in my head that didn't update the Gemfiles, but looking again I think it does. Yes, definitely think it makes sense to add a note :smile_cat:

emdupre replied:
@jasmainak I think this is resolved with the new note about updating the books -- let me know if you think it should be clearer !

jasmainak commented on jupyter_book/book_template/Dockerfile:
```diff
@@ -0,0 +1,5 @@
+COPY Gemfile .
+RUN apk --update add make gcc g++ libc-dev
+RUN gem install bundler
+RUN bundle install
```
There is one other potential issue with this. How will I use jupyter-book with mybinder? Because mybinder expects a requirements.txt or Dockerfile.

emdupre replied:
What might be ideal is to move the Dockerfile itself up a directory (or two) and encourage users to pull the image down from Dockerhub. Does Jupyter have a Dockerhub account, @choldgraf ?

choldgraf replied:
hmmm, that is a good question...@carreau any chance you know about the docker images and jupyter?

emdupre replied:
Currently hosted on my personal DockerHub account. Happy to move it, though :)

jasmainak commented:
I've been using this for the last few days (thanks @emdupre !). It works nicely on my laptop. However, I discovered one other limitation of this approach that one needs to be aware. If you don't have sudo (e.g., on work computers), you can't install docker (and it's likely very difficult to convince some IT departments :). Conda on the other hand works also without root access.

emdupre commented:
In that case you'd build a Singularity container :) [From dockerhub](https://www.sylabs.io/guides/2.6/user-guide/build_a_container.html#downloading-a-existing-container-from-docker-hub) (if we host it there) or [by building a Singularity file directly from the Dockerfile](https://www.sylabs.io/guides/2.6/user-guide/build_a_container.html#building-containers-from-singularity-recipe-files).

I'll go ahead and add docs for doing either, but not sure if we should bet on this horse or on https://github.com/jupyter/jupyter-book/issues/83#issuecomment-471982215 !

jasmainak commented:
okay I will experiment with Singularity. But I don't want to hold up this PR because of that. Feel free to move on once the `Dockerfile` issue is solved !

mathieuboudreau commented on jupyter_book/book_template/content/guide/03_build.md:
```diff
@@ -42,16 +42,52 @@ From here, you have **two options**
+   -v /PATH/TO/YOUR/CONTENT:/srv/jekyll \
+   -p 4000:4000 \
+   -it -u 1000:1000 \
+   jupyter-book bundle exec jekyll serve --host 0.0.0.0
```
I followed these instructions for a Docker build, but for some reason I am encountering a ruby error when executing the `docker run` line :/ 

See screenshots of my terminal here: https://imgur.com/a/MLDSC9X

Did I miss a step, or misunderstand the instructions?

emdupre replied:
I haven't seen that exact error, but just as a guess -- can you delete your local `Gemfile.lock` before running the command ?

mathieuboudreau replied:
Ah ha, yes that worked! I saw the messages above about it, but I assumed because they were running older builds of jupyter-book, that deleting it might not be needed for this branch. 

Thanks a lot @emdupre - this feature is **amazing** 🎉🎉🎉🎉 🚀 🚀 🚀 !

choldgraf commented:
@emdupre is this one ready to go?

mathieuboudreau commented:
Any advice for how to integrate this into a project that depends on another Dockerfile for the Jupyter Notebooks running on MyBinder? And if so would it be worth documenting here?

e.g. Since in my project, https://github.com/qMRLab/t1_book, I already have a Dockerfile in the root of the repo to build the Jupyter environment, and I think MyBinder only looks for dockerfiles at this repo level, what's the best strategy to manage both of these for a live site on github? Would it be to make my old site in a new directory (somewhat like what's done currently in this repo here), keep my Jupyter-Dockerfile in root and add your Ruby-Dockerfile in the new directory? Or make the live site from a gh-pages branch with the Ruby-Dockerfile there, and keep my Jupyter-Dockerfile in master or another branch? SOrry if this is confusing; it's the first time I have a repo needing two Dockerfiles, and the location you placed yours (the demo book) is where my old dockerfile lived.

emdupre commented:
Not yet, but I can finish it tonight -- for now I can host the image on my
personal dockerhub account ?

On Thu., Mar. 14, 2019, 17:25 Chris Holdgraf, <notifications@github.com>
wrote:

> @emdupre <https://github.com/emdupre> is this one ready to go?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyter/jupyter-book/pull/127#issuecomment-473069994>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AOUk5-zNoLM4zSW0bISrsjweM0loExxlks5vWr5JgaJpZM4bmdCe>
> .
>

emdupre commented:
@mathieuboudreau, I've just updated this PR and pointed to my DockerHub image -- can you try that by chance ? I've also moved the Dockerfile up one directory so it shouldn't conflict with project specific Dockerfiles !

PS @jasmainak: If you're still interested in using Singularity I updated the build instructions to explain how to do it !

mathieuboudreau commented:
@emdupre pulling from Dockerhub works great - thanks a lot! 💯

emdupre commented:
OK, I think it's ready now @choldgraf if you want to do a final review ! :smile_cat:

choldgraf commented:
this all looks great to me, and given that a few folks have already said this works for them, I think we should merge! 🎉 thanks @emdupre :-)

choldgraf commented:
@emdupre two quick things:

1. I don't see this documentation reflected in the jupyter book online docs...I wonder if we've got an error in the circle build somewhere?

2. I tried running this on a windows machine and got this error:

   

```
Configuration file: none
            Source: /srv/jekyll
       Destination: /srv/jekyll/_site
 Incremental build: disabled. Enable with --incremental
      Generating...
       Jekyll Feed: Generating feed for posts
   GitHub Metadata: No GitHub API authentication could be found. Some fields may be missing or have incorrect data.
   GitHub Metadata: Error processing value 'title':
  Liquid Exception: No repo name found. Specify using PAGES_REPO_NWO environment variables, 'repository' in your configuration, or set up an 'origin' git remote pointing to your github.com repository. in feed.xml
             ERROR: YOUR SITE COULD NOT BE BUILT:
                    ------------------------------------
                    No repo name found. Specify using PAGES_REPO_NWO environment variables, 'repository' in your configuration, or set up an 'origin' git remote pointing to your github.com repository.
```

is that something you've run into before?

emdupre commented:
I've never seen that :thinking:  -- but could you try this ? https://github.com/github/pages-gem/issues/399#issuecomment-301827749

:crossed_fingers: :crossed_fingers: it works !!
